managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java

```diff
@@ -276,14 +280,54 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
     private static final class ReadEntriesCallbackWithContext {
         final ReadEntriesCallback callback;
         final Object ctx;
+        final long startEntry;
+        final long endEntry;
     }
+
+    private CachedPendingRead findBestCandidate(PendingReadKey key, Map<PendingReadKey, CachedPendingRead> ledgerCache,
+                                                AtomicBoolean created) {
+        synchronized (ledgerCache) {
+            CachedPendingRead existing = ledgerCache.get(key);
+            if (existing != null) {
+                return existing;
+            }
+            // TODO: use a sorted map
+            for (Map.Entry<PendingReadKey, CachedPendingRead> entry : ledgerCache.entrySet()) {
+                if (entry.getKey().includes(key)) {
+                    //log.info("partial match {} with {}", key, entry.getKey());
+                    return entry.getValue();
+                }
+            }
+            created.set(true);
+            CachedPendingRead newRead = new CachedPendingRead(key, ledgerCache);
+            ledgerCache.put(key, newRead);
+            return newRead;
+        }
```

Comment:
Please add Prometheus metrics so that we could follow the effectiveness of this:
- pending read direct cache hit
- pending read includes cache hit
- pending read cache miss

Since the next step could be to add support for de-duplicating overlapping ranges, it would be useful to add metrics for this before implementing it.
One possible metric would be to add a counter for "pending read cache miss overlapping range" and calculate an average metric of the size of the overlapping range.

Reply:
done

Reply:
I didn't do this part "calculate an average metric of the size of the overlapping range."